lib/ruby_lsp/requests/on_type_formatting.rb

```diff
@@ -0,0 +1,132 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # ![On type formatting demo](../../misc/on_type_formatting.gif)
+    #
+    # The [on type formatting](https://microsoft.github.io/language-server-protocol/specification#textDocument_onTypeFormatting)
+    # request formats code as the user is typing. For example, automatically adding `end` to class definitions.
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Foo # <-- upon adding a line break, on type formatting is triggered
+    #   # <-- cursor ends up here
+    # end # <-- end is automatically added
+    # ```
+    class OnTypeFormatting < BaseRequest
+      extend T::Sig
+
+      END_REGEXES = T.let([
+        /(if|unless|for|while|class|module|until|def|case).*/,
+        /.*\sdo/,
+      ], T::Array[Regexp])
+
+      sig { params(document: Document, position: Document::PositionShape, trigger_character: String).void }
+      def initialize(document, position, trigger_character)
+        super(document)
+
+        scanner = Document::Scanner.new(document.source)
+        line_begin = position[:line] == 0 ? 0 : scanner.find_position({ line: position[:line] - 1, character: 0 })
+        line_end = scanner.find_position(position)
+        line = T.must(@document.source[line_begin..line_end])
+
+        @indentation = T.let(find_indentation(line), Integer)
+        @previous_line = T.let(line.strip.chomp, String)
+        @position = position
+        @edits = T.let([], T::Array[LanguageServer::Protocol::Interface::TextEdit])
```

Comment:
I think we can remove the namespaces as we've added shortcut [here](https://github.com/Shopify/ruby-lsp/blob/94d7bda937e2609d3e450e75c2c4f51e4e040389/lib/ruby_lsp/handler.rb#L9)?